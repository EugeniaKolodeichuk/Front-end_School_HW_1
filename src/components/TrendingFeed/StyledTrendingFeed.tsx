import { Link } from 'react-router-dom';
import styled from 'styled-components';

export const StyledListNoorder = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-evenly;
  align-items: center;
  flex-direction: column;
  margin-top: 15px;
  margin-bottom: 1.5em;
  padding: 0.6em;
  background: #fefefe8c;
  color: #231f20;
  width: 350px;
  border-radius: 2px;
  box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.2),
    0px 1px 1px 0px rgba(0, 0, 0, 0.14), 0px 2px 1px 0px rgba(0, 0, 0, 0.12);
`;

export const StyledLink = styled(Link)`
  display: inline-block;
  text-decoration: none;
  padding: 10px;
  color: inherit;

  &:hover {
    color: #005eaa;
  }
`;

export const StyledFeedInfo = styled.div`
  display: inline-flex;
  align-items: center;
  flex-direction: row;
  margin-top: 15px;
`;
