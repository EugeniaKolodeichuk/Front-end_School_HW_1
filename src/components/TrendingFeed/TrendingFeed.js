import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Avatar from '../Avatar/Avatar';
import Stats from '../Stats/Stats';
import Video from '../Video/Video';
import styles from './TrendingFeed.module.css';

const TrendingFeed = ({ feed }) => {
  return (
    <div className={styles.list_noorder} key={feed.id}>
      <Video playAddress={feed.video.downloadAddr} />
      <p>{feed.author.signature}</p>
      <div className={styles.feed_info}>
        <Avatar
          avatarThumb={feed.author.avatarThumb}
          nickname={feed.author.nickname}
        />
        <p>
          <Link className={styles.link} to={`/profile/${feed.author.uniqueId}`}>
            <b>{feed.author.nickname}</b>
          </Link>
        </p>
      </div>
      <Stats
        commentCount={feed.stats.commentCount}
        diggCount={feed.stats.diggCount}
      />
    </div>
  );
};

TrendingFeed.propTypes = {
  feed: PropTypes.any.isRequired,
};
export default TrendingFeed;
