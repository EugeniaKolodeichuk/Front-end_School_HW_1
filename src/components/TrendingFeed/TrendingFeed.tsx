import React from 'react';
import { Link } from 'react-router-dom';
import Avatar from '../Avatar/Avatar';
import Stats from '../NewsStats/NewsStats';
import Video from '../Video/Video';
import styles from './TrendingFeed.module.css';
import Loader from '../Loader/Loader';

interface ITrendingFeed {
  feed: {
    video: {
      downloadAddr: string;
    };
    author: {
      signature: string;
      avatarThumb: string;
      nickname: string;
      uniqueId: string;
    };
    stats: {
      commentCount: number;
      diggCount: number;
    };
  };
}

const TrendingFeed = ({ feed }: ITrendingFeed) => {
  return feed.video.downloadAddr ? (
    <div className={styles.list_noorder} key={feed.author.nickname}>
      <Video downloadAddr={feed.video.downloadAddr} />
      <p>{feed.author.signature}</p>
      <div className={styles.feed_info}>
        <Avatar
          avatarThumb={feed.author.avatarThumb}
          nickname={feed.author.nickname}
        />
        <p>
          <Link className={styles.link} to={`/profile/${feed.author.uniqueId}`}>
            <b>{feed.author.nickname}</b>
          </Link>
        </p>
      </div>
      <Stats
        commentCount={feed.stats.commentCount}
        diggCount={feed.stats.diggCount}
      />
    </div>
  ) : (
    <Loader />
  );
};

export default TrendingFeed;
