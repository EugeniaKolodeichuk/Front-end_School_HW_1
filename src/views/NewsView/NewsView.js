import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import defaultImage from '../../images/default.jpg';
import styles from './NewsView.module.css';
import Loader from '../../components/Loader/Loader';
import { getTrendingFeed } from '../../service/app';

export default function NewsView() {
  const [trendingFeed, setTrendingFeed] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [fetching, setFetching] = useState(true);

  const scrollHandler = event => {
    if (
      event.target.documentElement.scrollHeight -
        (event.target.documentElement.scrollTop + window.innerHeight) <
      100
    ) {
      setFetching(true);
    }
  };

  useEffect(() => {
    document.addEventListener('scroll', scrollHandler);

    return function () {
      document.removeEventListener('scroll', scrollHandler);
    };
  }, []);

  useEffect(() => {
    try {
      getTrendingFeed(currentPage).then(data => {
        setTrendingFeed([...trendingFeed, ...data]);
        setCurrentPage(previousState => previousState + 1);
      });
    } catch (error) {
      console.error(error);
    } finally {
      setFetching(false);
    }
  }, [fetching]);

  return (
    <div className={styles.wrap}>
      {trendingFeed.length > 0 ? (
        trendingFeed.map(user => (
          <div className={styles.list_noorder} key={user.id}>
            {user.video.playAddr ? (
              <video width="300px" controls="controls">
                <source src={user.video.playAddr} type="video/mp4;" />
                <track kind="captions" srcLang="en" label="english_captions" />
              </video>
            ) : (
              <Loader />
            )}
            <p>{user.author.signature}</p>
            <div className={styles.user_info}>
              {user.author.avatarThumb ? (
                <img
                  key={user.video.id}
                  width="70px"
                  src={user.author.avatarThumb}
                  alt={user.author.nickname}
                  className={styles.avatar}
                />
              ) : (
                <img
                  width="150px"
                  src={defaultImage}
                  alt={user.author.nickname}
                />
              )}
              <p>
                <Link
                  className={styles.link}
                  to={`/profile/${user.author.uniqueId}`}
                >
                  <b>{user.author.nickname}</b>
                </Link>
              </p>
            </div>
            <p>
              Comments: <b>{user.stats.commentCount}</b>
            </p>
            <p>
              Likes: <b>{user.stats.diggCount}</b>
            </p>
          </div>
        ))
      ) : (
        <Loader />
      )}
    </div>
  );
}
