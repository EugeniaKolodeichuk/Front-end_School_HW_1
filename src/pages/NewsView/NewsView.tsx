import React, { useState, useEffect } from 'react';
import shortid from 'shortid';
import styles from './NewsView.module.css';
import Loader from '../../components/Loader/Loader';
import { getTrendingFeed } from '../../service/app';
import TrendingFeed from '../../components/TrendingFeed/TrendingFeed';
import { ITrendingFeed } from '../../types/feedTypes';

export default function NewsView() {
  const [trendingFeed, setTrendingFeed] = useState<ITrendingFeed[]>([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [fetching, setFetching] = useState(true);

  const scrollHandler = (event: any): void => {
    if (
      event.target.documentElement.scrollHeight -
        (event.target.documentElement.scrollTop + window.innerHeight) <
      100
    ) {
      setFetching(true);
    }
  };

  useEffect(() => {
    document.addEventListener('scroll', scrollHandler);

    return function () {
      document.removeEventListener('scroll', scrollHandler);
    };
  }, []);

  useEffect(() => {
    try {
      getTrendingFeed(currentPage).then(data => {
        setTrendingFeed([...trendingFeed, ...data]);
        setCurrentPage(previousState => previousState + 1);
      });
    } catch (error) {
      console.error(error);
    } finally {
      setFetching(false);
    }
  }, [fetching]);

  return (
    <div className={styles.wrap} key={shortid.generate()}>
      {trendingFeed.length > 0 ? (
        trendingFeed.map(feed => (
          <TrendingFeed key={shortid.generate()} feed={feed} />
        ))
      ) : (
        <Loader />
      )}
    </div>
  );
}
