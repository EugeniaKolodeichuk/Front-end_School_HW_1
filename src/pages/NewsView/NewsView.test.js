import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import '@testing-library/jest-dom';
import { render, screen, act } from '@testing-library/react';
import NewsView from './NewsView';
import axios from 'axios';
import TrendingFeed from '../../components/TrendingFeed/TrendingFeed';

jest.mock('axios');
jest.mock('../../components/TrendingFeed/TrendingFeed', () => {
  return {
    __esmodule: true,
    default: () => {
      return <p>Mocked component</p>;
    },
  };
});

let container = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});
describe('#NewsView.js', () => {
  it('render news page', () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <NewsView />
      </Router>,
      container,
    );
    expect(container).toMatchSnapshot();
  });

  it('should have the correct content', async () => {
    axios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: 'result-test' }),
    );
    render(<NewsView />);
    /* expect(axios.create).toBeCalled(); */
    screen.debug();
  });
});
