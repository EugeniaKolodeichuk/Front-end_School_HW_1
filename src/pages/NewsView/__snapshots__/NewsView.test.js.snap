// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#NewsView.js render news page 1`] = `<div />`;
